extensions/artemis-core/runtime/src/main/java/io/quarkus/artemis/core/runtime/ArtemisBuildConfig.java

```diff
@@ -0,0 +1,35 @@
+package io.quarkus.artemis.core.runtime;
+
+import java.util.Optional;
+
+import io.quarkus.runtime.annotations.ConfigItem;
+import io.quarkus.runtime.annotations.ConfigPhase;
+import io.quarkus.runtime.annotations.ConfigRoot;
+
+@ConfigRoot(name = "artemis", phase = ConfigPhase.BUILD_AND_RUN_TIME_FIXED)
+public class ArtemisBuildConfig {
+
+    public enum Protocol {
+        CORE,
+        JMS
+    }
+
+    /**
+     * Artemis protocol, defaults to JMS if it is on classpath else to CORE
+     */
+    @ConfigItem
+    public Optional<Protocol> protocol;
+
+    public Protocol getProtocol() {
+        return protocol.orElseGet(ArtemisBuildConfig::getDefaultProtocol);
+    }
+
+    public static Protocol getDefaultProtocol() {
+        try {
+            Class.forName("io.quarkus.artemis.jms.runtime.ArtemisJmsTemplate");
```

Comment:
Not a big fan of these runtime discovery in Quarkus. Any reasons why the JMS extension could not send a build step that can be listened to by artemis core to decide at build time instead of runtime?

Reply:
I am not a big fan myself either (although it is only being used build time). Have you a good example in the code how to do this? I will add something. And would you ditch the option? I was doubting, because I think if you include artemis-jms (which depends on artemis-core as the JMS client is a wrapper around their core client) you probably want to use the JMS client.

Reply:
Look at this. Agroal is the connection pool that publishes a datasource driver

https://github.com/quarkusio/quarkus/blob/master/extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java#L71
https://github.com/quarkusio/quarkus/blob/master/extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java#L104

And some listen to these build items.
https://github.com/quarkusio/quarkus/blob/master/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java#L120

Reply:
Thanks, I have added a commit that uses a BuildItem instead of `Class.forName`. I got rid of the BuildConfig, as I don't see any scenario where you going to include `artemis-jms` but still want to use the Core API.